### PR TITLE
xeus-cling: init at 0.15.3

### PR DESCRIPTION
**Pull request by thomasjm**
###### Description of changes

The Jupyter kernel for C++ based on Cling.

~~This is a draft until I can get https://github.com/NixOS/nixpkgs/pull/247253 and https://github.com/NixOS/nixpkgs/pull/245340 merged, and then it will be ready to go.~~

There are currently 2 patches included with this PR. One has already been merged upstream and the other is in progress, so hopefully we'll have a new release soon incorporating both and then we can remove the patches.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thomasjm** [comment]
Hi @tjni, I was wondering if you might be willing to review this?

File: pkgs/applications/editors/jupyter-kernels/xeus-cling/xeus-cling.nix [hunk 1]
```diff
@@ -0,0 +1,73 @@
+{ lib
+, callPackage
+, cmake
+, fetchFromGitHub
+, gcc
+, git
+
+, clangStdenv
+, llvmPackages_9
+
```
Comment (GTrunSec) [inline pkgs/applications/editors/jupyter-kernels/xeus-cling/xeus-cling.nix:10]: Could you remove the unnecessary whitespace from the code? We need to maintain the same style as other packages, like https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/pp/ppsspp/package.nix#L20

Comment (thomasjm) [reply pkgs/applications/editors/jupyter-kernels/xeus-cling/xeus-cling.nix]: Done

**Comment by GTrunSec** [comment]
@thomasjm 
Did you test  the `nix build .\#xeus-cling` locally?
```
error: builder for '/nix/store/md5jjwjw2wj04699m3kdpbdg4gj3i4qb-xeus-cling-0.15.3.drv' failed with exit code 2;
       last 10 log lines:
       > /build/source/include/xeus-cling/xoptions.hpp:21:24: note: copy constructor of 'argparser' is implicitly deleted
because base class 'argparse::ArgumentParser' has a deleted copy constructor
       >     struct argparser : public argparse::ArgumentParser
       >                        ^
       > /nix/store/2b054ars50sifp1b5lr7dw1s33iawizq-argparse-3.0/include/argparse/argparse.hpp:1477:3: note: 'ArgumentPar
ser' has been explicitly marked deleted here
       >   ArgumentParser(const ArgumentParser &other) = delete;
       >   ^
       > 1 error generated.
       > make[2]: *** [CMakeFiles/xeus-cling.dir/build.make:146: CMakeFiles/xeus-cling.dir/src/xmagics/executable.cpp.o] E
rror 1
       > make[1]: *** [CMakeFiles/Makefile2:166: CMakeFiles/xeus-cling.dir/all] Error 2
       > make: *** [Makefile:136: all] Error 2
       For full logs, run 'nix log /nix/store/md5jjwjw2wj04699m3kdpbdg4gj3i4qb-xeus-cling-0.15.3.drv'.
```

**Comment by thomasjm** [comment]
No, I just rebased on `master`. The problem is that Nixpkgs now has version 3 of the `argparse` dependency, while this package needs ~2.9.

I added a commit to accomplish this, but I'm not sure if this is how it's normally done.

**Comment by GTrunSec** [comment]
nixpkgs-review pr 244777
<details>
git worktree add /home/guangtao/.cache/nixpkgs-review/pr-244777-1/nixpkgs aba830385baac22d06a22085f5b5baeb88c8[15/19940]
Preparing worktree (detached HEAD aba830385baa)
Updating files: 100% (38022/38022), done.
HEAD is now at aba830385baa Merge pull request #267782 from kirillrdy/ruby_3_3
$ nix-env --extra-experimental-features no-url-literals --option system x86_64-linux -f <nixpkgs> --nix-path nixpkgs=/home
/guangtao/.cache/nixpkgs-review/pr-244777-1/nixpkgs nixpkgs-overlays=/run/user/1000/tmplv9p4bk0 -qaP --xml --out-path --sh
ow-trace --no-allow-import-from-derivation
$ git merge --no-commit --no-ff 86262c024073fba280bcb377edc47f6c4bdbb288
Automatic merge went well; stopped before committing as requested
$ nix-env --extra-experimental-features no-url-literals --option system x86_64-linux -f <nixpkgs> --nix-path nixpkgs=/home
/guangtao/.cache/nixpkgs-review/pr-244777-1/nixpkgs nixpkgs-overlays=/run/user/1000/tmplv9p4bk0 -qaP --xml --out-path --sh
ow-trace --no-allow-import-from-derivation --meta
1 package added:
xeus-cling (init at 0.15.3)

$ nix build --nix-path nixpkgs=/home/guangtao/.cache/nixpkgs-review/pr-244777-1/nixpkgs nixpkgs-overlays=/run/user/1000/tm
plv9p4bk0 --extra-experimental-features nix-command no-url-literals --no-link --keep-going --no-allow-import-from-derivati
on --option build-use-sandbox relaxed -f /home/guangtao/.cache/nixpkgs-review/pr-244777-1/build.nix

Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/244777

1 package built:
xeus-cling

$ /nix/store/vxx4c6gc2zgfw870b40f06dmli6ljp34-nix-2.17.0/bin/nix-shell /home/guangtao/.cache/nixpkgs-review/pr-244777-1/sh
ell.nix --nix-path nixpkgs=/home/guangtao/.cache/nixpkgs-review/pr-244777-1/nixpkgs nixpkgs-overlays=/run/user/1000/tmplv9
p4bk0
</details>

@thomasjm, Could you reset your changes to one commit? thanks

**Comment by thomasjm** [comment]
Done

**Review by GTrunSec — APPROVED** [review]

**Comment by thomasjm** [comment]
Pinging @teto here as well, this should be ready to merge!

**Comment by GTrunSec** [comment]
@natsukium ready to merge, thanks

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -17579,6 +17579,10 @@ with pkgs;
     jre = jre8;
   };
 
+  inherit (callPackage ../applications/editors/jupyter-kernels/xeus-cling { })
+    cpp11-kernel cpp14-kernel cpp17-kernel cpp2a-kernel;
+  xeus-cling = callPackage ../applications/editors/jupyter-kernels/xeus-cling/xeus-cling.nix { };
```
Comment (teto) [inline pkgs/top-level/all-packages.nix:17584]: same remark as https://github.com/NixOS/nixpkgs/pull/171202/files#r1397979118

**Comment by teto** [comment]
> here are currently 2 patches included with this PR. One has already been merged upstream and the other is in progress, so hopefully we'll have a new release soon incorporating both and then we can remove the patches.

no release since then ?

**Comment by thomasjm** [comment]
> no release since then ?

I once waited 33 months for the maintainers to land a 10-line change :)

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (teto) [reply pkgs/top-level/all-packages.nix:17584]: I am annoyed to see those kernels at the top-level still since we dont make much of it without jupyter. Would it make sense to have the kernels in a package set like luaPackages/pythonPackages -> jupyterKernels = { xeus-cling , coq_kernel, ihaskell ... }. Is that on your roadmap. What do you think ?

Comment (thomasjm) [reply pkgs/top-level/all-packages.nix:17584]: Yeah, I did argue just now against exposing kernels at the top-level :)

It's definitely on my roadmap. I could either do it in this PR or a subsequent one. Maybe a subsequent one would be cleaner? But I'm up for whatever you prefer.
